### PR TITLE
Fix output for "Printing Dependent Evars Line"

### LINKED ISSUE #4504: dependent evar printout in 8.5rc1 insufficient
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4504
From: @hendriktews
Reported version: 8.5
CC:   @HuStmpHrrr

### LINKED ISSUE #10399: dependent evars line empty
In Coq 8.9.1, when I feed

    Set Printing Dependent Evars Line.
    Lemma x : exists(i : nat), i = 0.
    Proof using.
      eexists.

into `coqtop -emacs`, the dependent evars line shows one open existential:

    (dependent evars: ?X6 open,)

When I do the same on the current git master (8acc04d4c33ed41ac8dfa668a026010d4592388d), I only see

    (dependent evars:)

Was that change intentional?

(Some background: The dependent evar line was introduced by Arnaud Spiwack in 2011 for Coq 8.4 on my request for prooftree, the proof tree visualization program for Proof General and Coq. For 8.4 and 8.5 the line enables prooftree to highlight commands that introduce or instantiate a particular existential variable. Since 8.6 the feature is broken, because there is no way to link the evars appearing in the goals with those in the dependent evar line.)

### LINKED ISSUE #10420: Incomplete evars info when 'Printing Dependent Evars Line' enabled
# Feature request

Please re-enable and extend the dependent evars line, such that it shows 

1. all not instantiated evars
2. all instantiated evars and for each of them the evars appearing in the instantiating term
3. the mapping between internal evar names and their printed representation
4. all the (open) evars apearing in the current goal

For the example proof script appended below, after `apply modpon.`, the dependent evar line should look similar to

    (dependent evars: ?X4 open, ?X5 using ?X11 ?X10, ?X10 using ?X11, ?X11 open
     this goal: ?X4, ?X11
     mapping: ?P : ?X4, ?Goal2 : ?X11)

Here line 1 (realizing point 1 and 2) is identical to the dependent evars line that is printed in Coq 8.4 till 8.9 and in versions before commit 6608f64f001f8f1a50b2dc41fefdf63c0b84b270. Since that commit the dependent evars line is always empty on the master branch, see also #10399.

Point 3 (realized by line 3) is a necessary enhancement because since 8.5 there is no way to connect the evars in the proofs with the identifiers appearing in the dependent evar line.

Instead of point 4 (realized by line 2) one could extract all matches for `\?[a-z0-9]+` from the goal text. But if it is not too much effort, an explicit list inside the dependent evars line seems more robust.

The example output above means that
- evars `?X4` and `?X11` are not instantiated
-`?X5` is instantiated with a term containing `?X11` and `?X10`
-`?X10` is instantiated with a term containing `?X11`
- the current goal contains `?X4` and `?X11`
- `?X4` is printed as `?P` and `?X11` as `?Goal2`

The dependent evars line should only be printed when the flag Printing Dependent Evars Line is on. It should be printed for the current goal and for the command `Show Goal 9 at <state>` from PR #10356.

**Note**  The dependent evars output appears only in coqtop (not CoqIDE).

# Background

If this feature request is granted, prooftree (the proof visualization program for Coq/Proof General) can support the following features.
- closed branches of the proof are shown in a different color, if they still depend on not yet instantiated evars
- the evar dialog shows the dependencies among evars and can highlight proof commands that introduced or instantiated a specific evar

These features were fully working for Coq 8.4 with a dependent evar line realizing points 1 and 2 that was implemented by Arnaud Spiwack in 2011. Since Coq 8.5, however, they are more or less broken, because or the missing link between different representations of the same evar.

# Example proof script

    Set Printing Dependent Evars Line.
    Lemma strange_imp_trans : 
      forall P Q R : Prop, (Q -> R) -> (P -> Q) -> (P -> Q) -> P -> R.
    Proof.
      auto.
    Qed.

    Lemma modpon : forall P Q : Prop, (P -> Q) /\ P -> Q.
    Proof.
      tauto.
    Qed.

    Section eex.
      Variables P1 P2 P3 P4 : Prop.
      Hypothesis p12 : P1 -> P2.
      Hypothesis p123 : (P1 -> P2) -> P3.
      Hypothesis p34 : P3 -> P4.

      Lemma p14 : P4.
      Proof.
        eapply strange_imp_trans. 
              apply modpon.
            intros H.
            split.
              eexact p34.
            apply p123.
            eexact H.
          intros H.
          split.
            trivial.
          apply p123.
          trivial.
        trivial.
      Qed.
    End eex.

### PR DESCRIPTION
**Pull request by jfehrle**
(Replacing #10483 which incorrectly had the branch in the master repository.)

Commit 6608f64f001f8f1a50b2dc41fefdf63c0b84b270 by @herbelin broke the output triggered by the `Printing Dependent Evars Line` flag.  The first commit here addresses this, fixing #10399.  There may be a better/simpler solution.  I don't know the purpose of that commit, so (very conservatively) I created a new routine `Evd.evars_of_term2`.  The routines that call this new routine are only called from `print_dependent_evars`, with `seeds` coming from `pr_open_subgoals_diff`, so I think that's safe.

The second commit addresses #4504 and #10420, a fix/enhancement adding a mapping between internal evar names and user-visible evar names.

Sample output: 

```
(dependent evars: ?X4 open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 open;
mapping: ?X4 : ?P, ?X7 : ?Goal, ?X8 : ?Goal0, ?X9 : ?Goal1, ?X11 : ?Goal2)
```

Fixes / closes #4504
Fixes / closes #10399
Fixes / closes #10420

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by jfehrle** [comment]
@herbelin Is there an easy way to get the list of evars that are visible in the current goal?

**Comment by jfehrle** [comment]
Updated:
- Added printing of dependent evar info to `Show Goal <gid> at <sid>.`, which will show only dependent evars reachable from <gid>.
- Added filtering of the `mapping:` output so it only shows items that appear in `dependent evars:`

I think that addresses my earlier question "Is there an easy way to get the list of evars that are visible in the current goal?" but I'd appreciate comments from the reviewers.

Hendrik, please try this, test it more exhaustively and let me know if this is good.  We won't merge this until you have prooftree working with this.

Sample output:

```
<prompt>p14 < 18 |p14| 0 < </prompt>          apply modpon.
3 focused subgoals
(shelved: 2) (ID 7)

  P1, P2, P3, P4 : Prop
  p12 : P1 -> P2
  p123 : (P1 -> P2) -> P3
  p34 : P3 -> P4
  ============================
  ?P -> (?Goal2 -> P4) /\ ?Goal2

subgoal 2 (ID 8) is:
 ?P -> (?Goal2 -> P4) /\ ?Goal2
subgoal 3 (ID 9) is:
 ?P

(dependent evars: ?X4 open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 open;
mapping: ?X4 : ?P, ?X11 : ?Goal2)

<prompt>p14 < 19 |p14| 0 < </prompt>
Show Goal 7 at 19.
goal ID 7 at state 19

  P1, P2, P3, P4 : Prop
  p12 : P1 -> P2
  p123 : (P1 -> P2) -> P3
  p34 : P3 -> P4
  ============================
  ?P -> (?Goal2 -> P4) /\ ?Goal2

(dependent evars: ?X4 open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 open;
mapping: ?X4 : ?P, ?X11 : ?Goal2)

<prompt>p14 < 19 |p14| 0 < </prompt>Show Goal 8 at 19.
goal ID 8 at state 19

  P1, P2, P3, P4 : Prop
  p12 : P1 -> P2
  p123 : (P1 -> P2) -> P3
  p34 : P3 -> P4
  ============================
  ?P -> (?Goal2 -> P4) /\ ?Goal2

(dependent evars: ?X4 open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 open;
mapping: ?X4 : ?P, ?X11 : ?Goal2)

<prompt>p14 < 19 |p14| 0 < </prompt>Show Goal 9 at 19.
goal ID 9 at state 19

  P1, P2, P3, P4 : Prop
  p12 : P1 -> P2
  p123 : (P1 -> P2) -> P3
  p34 : P3 -> P4
  ============================
  ?P

(dependent evars: ?X4 open;
mapping: ?X4 : ?P)
```

**Comment by herbelin** [comment]
> Commit 6608f64 by @herbelin broke the output triggered by the Printing Dependent Evars Line flag. The first commit here addresses this, fixing #10399. There may be a better/simpler solution. I don't know the purpose of that commit, so (very conservatively) I created a new routine Evd.evars_of_term2

Hi @jfehrle, thanks a lot for having investigated. I think you don't need to make the code more complicated by adding in parallel an `evars_of_term2`. There is a naive bug in 6608f64 coming from an overlook of mine in the code of `evars_of_term`.  It should be changed as follows:
```
@@ -1403,7 +1403,8 @@ end
 
 let evars_of_term evd c =
   let rec evrec acc c =
-    match MiniEConstr.kind evd c with
+    let c = MiniEConstr.whd_evar evd c in
+    match kind c with
     | Evar (n, l) -> Evar.Set.add n (Array.fold_left evrec acc l)
     | _ -> Constr.fold evrec acc c
   in
```

> @herbelin Is there an easy way to get the list of evars that are visible in the current goal?

If ever this question is still actual, isn't `top_evars` and `gather_dependent_evars` what you are looking for?

**Comment by jfehrle** [comment]
@herbelin Thanks for the change.  It is better but it doesn't have the "using" evars that were present in the old output and in this PR as-is.  Hendrik tells me this information is essential for his tool.  Perhaps there's an additional fix, or do we need to keep the separate routines?

Output from this PR as-is after "apply modpon" using proof in https://github.com/coq/coq/issues/10420:

```
(dependent evars: ?X4 open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 open;
mapping: ?X4 : ?P, ?X11 : ?Goal2)
```

Output with your suggested change:

```
(dependent evars: ?X4 open, ?X11 open;
mapping: ?X4 : ?P, ?X11 : ?Goal2)
```

> If ever this question is still actual, isn't top_evars and gather_dependent_evars what you are looking for?

Yes, I think so.

**Comment by herbelin** [comment]
Is your second commit not precisely intended to provide the `using` info? I.e. would my fix to `evars_of_term` replacing first commit + your second commit be ok?

**Comment by jfehrle** [comment]
No.  The second commit only adds the "mapping:" line and tweaks the positioning of the separators to be more standard.  The "using" info was already being shown.  `Printer.print_dependent_evars` still calls `Evarutil.gather_dependent_evars` to get the underlying info.

**Comment by herbelin** [comment]
OK, then there was a confusion on the different uses of `evars_of_term` (whether it is supposed to return the undefined evars or the evar nodes, independently on whether they are defined or not). So, yes, a variant of `evars_of_term` is needed to implement the `using` feature. A name such as `evar_nodes_of_term` or `literate_evars_of_term` might be clearer than `evars_of_term2`. Thanks.

**Comment by jfehrle** [comment]
OK, thanks, just pushed an update with your fix and renaming the routine.  Let's wait to hear back from Hendrik before merging in case there are any other details to address.

**Comment by herbelin** [comment]
As a side remark, this is an instance of a more general problem. We would need three types where we only have two:
- `econstr` as an abstraction of terms with undefined holes
- an evar-closed variant of `constr` as the type of terms without holes
- a low-level variant of `constr` as a representation of terms with evars seen as pointers

We use `constr` for the last two while we would need two types (compare `to_constr` and `unsafe_to_constr`). @ppedrot

**Review by Zimmi48 — APPROVED** [review]
Doc was reviewed already in the previous PR.

**Comment by hendriktews** [comment]
> * Added printing of dependent evar info to `Show Goal <gid> at <sid>.`, which will show only dependent evars reachable from <gid>.

Thanks! I tried the current version of this PR and it works fine so far. I noticed that the dependent evar info for `Show Goal <gid> at <sid>.` only shows the information for the existential variables of the current goal, while the dependent evar info, that is printed with the current goal after a proof command, shows all existential variables.

After giving it some more thought, I came to the conclusion, that for `Show Goal <gid> at <sid>.` I only need the mapping line. (For the dependent evar info printed after the current goal, I do need both lines, as it is implemented in this PR currently.) Therefore, the slight difference in the evar info for Show Goal is fine for me.

I have another small suggestion for improvement: When finishing a proof, Coq prints the dependent evar info directly concatenated to the no more subgoals message. For instance, after

    Set Printing Dependent Evars Line.
    Lemma simple : True.
    Proof using.
      trivial.

Coq prints

    No more subgoals.(dependent evars: ;mapping: )

Would it be possible to separate the message from the evar info with an empty line?

**Comment by hendriktews** [comment]
> Hendrik, please try this, test it more exhaustively and let me know if this is good.

Works fine also on other examples.

> We won't merge this until you have prooftree working with this.

I will take a while until I get prooftree running again. I would suggest to merge before that.

**Comment by jfehrle** [comment]
> I noticed that the dependent evar info for `Show Goal <gid> at <sid>.` only shows the information for the existential variables of the current goal, while the dependent evar info, that is printed with the current goal after a proof command, shows all existential variables.

My thought was that `Show Goal` would show only the evars that apply to that goal.  It's very easy to change but it sounds like it doesn't affect prooftree.

> I only need the mapping line.

I think it's better to always provide the two lines.  Prooftree can ignore the extra information.  This avoids adding conditionals to generate the output and to parse it on your side.  Changing this seems very prooftree specific; another tool may want all the info.

> Would it be possible to separate the message from the evar info with an empty line?

I'll take a look.  Sometimes these things are not so trivial.  For usability, PR/prooftree should filter out the `(dependent evars ... )` info and not display it.  It's not relevant to the user; it's just added visual clutter.

> I will take a while until I get prooftree running again. I would suggest to merge before that.

8.11 is several months away.  It's just as easy to keep the PR open and include any additional tweaks in it.  On the other hand, if merging it now will make your development easier, we could do that.  When do you think you may have prooftree working again?

**Comment by hendriktews** [comment]
> My thought was that `Show Goal` would show only the evars that apply to that goal.

OK, for prooftree it is fine as is.

> I think it's better to always provide the two lines.

I leave this decision to you. Either way is fine for me.

> For usability, PR/prooftree should filter out the `(dependent evars ... )` info and not display it.

Sure. I believe with an empty line the regular expressions for filtering would be a little bit simpler.

> On the other hand, if merging it now will make your development easier, we could do that.

It doesn't make a difference to me.

> When do you think you may have prooftree working again?

It's a hobby project, it can well take more than a month.

**Comment by jfehrle** [comment]
@herbelin Could you approve this if agreeable and perhaps do the merge?  Not sure if @ejgallego has any comments--it's a pretty routine change to coqloop.ml.  Thanks.

@hendriktews I added a blank line after `No more subgoals.` so it looks like this:

```
No more subgoals.

(dependent evars: ;
mapping: )
```

Your earlier comment had no line breaks at all, like this.  I assumed that was a just a typo:

```No more subgoals.(dependent evars: ;mapping: )```

Let me know when you have prooftree working.

**Comment by herbelin** [comment]
It was ok to me (currently, the rebase has a problem though - 56 commits).

**Comment by jfehrle** [comment]
I fixed the rebase.

File: printing/printer.ml
Comment (ejgallego) [inline printing/printer.ml]: Indentation seems incorrect here.

File: printing/printer.ml
Comment (ejgallego) [inline printing/printer.ml]: Fix indentation here.

File: toplevel/coqloop.ml
Comment (ejgallego) [inline toplevel/coqloop.ml]: indentation seems incorrect here.

File: toplevel/coqloop.ml
Comment (ejgallego) [inline toplevel/coqloop.ml]: Don't declare variables that you are not using, also modern OCaml prefers internal record qualification I think:
```
let { Proof.sigma } = Proof.data p in ...
```

**Review by ejgallego — CHANGES_REQUESTED** [review]
Looks good however the are some style problems to be ideally sorted out before merge.

Indentation in Coq is 2 spaces, and while not official (yet) we tend to consider ocp-indent default style as the reference. [we should make it official I know]

File: printing/printer.ml
Comment (jfehrle) [reply printing/printer.ml]: This has a tab to space conversion.  How many spaces to a tab character?

File: toplevel/coqloop.ml
Comment (jfehrle) [reply toplevel/coqloop.ml]: Like many things, it's not easy to find OCaml language details in the OCaml doc.

File: printing/printer.ml
Comment (jfehrle) [reply printing/printer.ml]: Looks like github thinks it's 6, which is a strange choice.

File: toplevel/coqloop.ml
Comment (ejgallego) [reply toplevel/coqloop.ml]: Indeed, coding style choices are mostly transmitted by folk knowledge.

**Comment by hendriktews** [comment]
> @hendriktews I added a blank line after `No more subgoals.`

Thanks, this looks fine.

> Your earlier comment had no line breaks at all, like this.

You are right, this was a copy-paste error.

> Let me know when you have prooftree working.

I will. I made some progress already, but it will still take some time.

File: toplevel/coqloop.ml
Comment (ejgallego) [reply toplevel/coqloop.ml]: By the way the construct is record pattern matching which is fairly standard, documented here: https://caml.inria.fr/pub/docs/manual-ocaml/patterns.html but yeah that's not very user friendly.

Real world OCaml does a bit better http://dev.realworldocaml.org/records.html#patterns-and-exhaustiveness

**Comment by hendriktews** [comment]
I came to the conclusion, that requirement 4 from the original feature request #10420 is not really optional. Requirement 4 was:

>     4. all the (open) evars apearing in the current goal

Originally I said that this requirement is optional, because I thought I can simply scan the sequent text for existential variables (with a suitable regular expression). However, I realized now, that this fails for existential variables occurring only in an implicit argument (which is not printed).

For the dependent evar line in the output of `Show Goal X at Y` requirement 4 is already fulfilled, but not for the dependent evar line in the default current goals output. I see the following options:

1. Ignore the problem of implicit arguments for now. It will probably take a while until this shows up in real examples. Prooftree for Coq 8.4 also relied on regular expressions and I never noticed that some existential variables are missing.
2. Issue a `Show Goal X at Y` always also for the current goal. It seems a bit odd to print each goal twice, but Proof General users wouldn't see this.
3. Add feature 4 to this PR.

I don't like option 2 too much, but I am fine with either 1 or 3. (When I started with #10315 I didn't expect that I would get all the features to make prooftree as good as it was within a few weeks. Therefore it is fine with me to leave something for the feature, but I wanted to tell you... And I apologize for the late changes of the feature request.)

File: toplevel/coqloop.ml
Comment (jfehrle) [reply toplevel/coqloop.ml]: Thanks.

Though if there's more than one field you would write:

```
  let Proof.{ sigma; goals } = Proof.data p in
```

or 

```
  let { Proof.sigma; Proof.goals } = Proof.data p in

```

as you did in Pfedit.get_nth_V82_goal?

**Comment by jfehrle** [comment]
@ejgallego updated to address your comments.  But l still want to consider @hendriktews latest comment.

**Review by ejgallego — APPROVED** [review]

File: toplevel/coqloop.ml
Comment (ejgallego) [reply toplevel/coqloop.ml]:
```
let { Proof.sigma; goals } = ... 
```
is a commonly accepted way, once OCaml resolves a record projection it can do resolution for the rest.

**Comment by jfehrle** [comment]
Hi Hendrik,

Let's see what's needed to do # 3.  Can you provide an example proof script using the current PR code showing the dependent evars output for some tactic plus the `Show Goal # at #` output?  With an explanation of how to compute what you need from this (your # 2 option)?  I'm supposing it's a simple set operation between the two outputs.

**Comment by hendriktews** [comment]
Jim,

I experimented with your patch and after a while I was able to add the code for what I need for prooftree. I created a pull request against your dependent_evars branch, please have a look. On the way I changed a number of other points, they are described in my commit message. I created the PR only for the discussion, I expect that it needs further changed and polishing and will then squashed together with your commit.

**Comment by jfehrle** [comment]
Can you provide an example proof script that illustrates the new behavior you requested (point 4)?  I will try  your earlier example with your updated code.

**Comment by hendriktews** [comment]
Please use the example proof script from #10420, but replace the line

```
          apply modpon.
```

with

```
          {
            apply modpon.
          }
```

After processing the first proof command, the `eapply strange_imp_trans.`, it will print

```
(dependent evars: ?X4 : ?P open, ?X5 : ?Q open, ; current goal: ?X5 )
```

At that point the whole proof has 2 evars, `?X4` printed as `?P` and `?X5` printed as `?Q`, but only `?X5` appears in the current goal.

After `apply modpon.` (but before processing the closing brace `}`), it will print

```
(dependent evars: ?X4 : ?P open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 : ?Goal2 open, ; no current goal)
```

The change in pr_subgoals, that causes `None` to be passed to `print_dependent_evars` in this situation, indirectly also causes the `no current goal` output.

After the closing brace, it prints

```
(dependent evars: ?X4 : ?P open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 : ?Goal2 open, ; current goal: ?X4 ?X5 ?X10 ?X11 )
```

Here the current goal contains `?P` and `?Goal2`. The output additionally also lists the instantiated evars that indirectly contain `?Goal2`.

You can also see, that I made the printout uglier, favoring a shorter `print_dependent_evars` function, but this is not meant for humans anyway.

**Comment by jfehrle** [comment]
> (dependent evars: ?X4 : ?P open, ?X5 using ?X10 ?X11, ?X10 using ?X11, ?X11 : ?Goal2 open, ; no current goal)

Where you have `using ?X10 ?X11` and `using ?X11`, you can figure out that `?X11` is `?Goal` from the other info?  Or you don't need the real name for `using`?

**Comment by hendriktews** [comment]
> Where you have `using ?X10 ?X11` and `using ?X11`, you can figure out that `?X11` is `?Goal` from the other info? Or you don't need the real name for `using`?

Instantiated evars are never printed, instead the instantiating term is printed. Therefore, I don't need the external name for instantiated evars.

**Comment by jfehrle** [comment]
OK.  Let's pick a better title for it, maybe "in current goal".  Do you need to distinguish between a current goal with no evars and when there is no currrent goal?

BTW, you can link directly to a github comment from the header of the comment:

image

**Comment by hendriktews** [comment]
> OK. Let's pick a better title for it, maybe "in current goal".

Fine with me. We could also leave the text completely out. 

> Do you need to distinguish between a current goal with no evars and when there is no currrent goal?

Prooftree distinguishes these cases, but I do not rely on the message in the dependent evar line. We could make the text uniform and just output no evars in the cases where there is no current goal.

> BTW, you can link directly to a github comment from the header of the comment:

Thanks for the hint, I was too lazy to search for this yesterday.

**Comment by jfehrle** [comment]
@hendriktews I incorporated your changes and tweaked the output format.  The test case files show the current output, though without showing the input :-(.

`Show Goal # at # ` may print different evars info than the default printing because it passes `seeds` of `[ gl ]` to `Printer.print_dependent_evars` whereas default printing passes `Proof.V82.top_evars p` (from `Printer.pr_open_subgoals_diff`).  Do you to keep them this way, or shall I pass `top_evars` in both calls?

**Comment by hendriktews** [comment]
> @hendriktews I incorporated your changes and tweaked the output format.

Thanks, this looks fine to me now.

> `Show Goal # at # ` may print different evars info [...] Do you to keep them this way, or shall I pass `top_evars` in both calls?

I don't have an opinion here, because from this command, I only need the `in current goal` information.

**Comment by jfehrle** [comment]
Great!  @ejgallego would you merge this when you get a chance?  The last changes were only in `print_dependent_evars`.  Thanks.

File: test-suite/output-coqtop/DependentEvars2.out
Comment (hendriktews) [inline test-suite/output-coqtop/DependentEvars2.out]: This file contains Windows line ends (CR + LF). Is that intended?

File: test-suite/output-coqtop/DependentEvars2.out
Comment (hendriktews) [inline test-suite/output-coqtop/DependentEvars2.out]: Is the test going to fail if the numbering of existential variables changes? And is that intended?

Comment (jfehrle) [reply test-suite/output-coqtop/DependentEvars2.out]: Yes, it will fail.  Not ideal, but making the test immune to such changes is non-trivial and not worth the trouble.  It's fairly easy to correct the test when the numbering changes.

File: test-suite/output-coqtop/DependentEvars2.out
Comment (jfehrle) [reply test-suite/output-coqtop/DependentEvars2.out]: I took them out, but it doesn't matter--the test runs fine either way.

**Comment by Zimmi48** [comment]
@jfehrle Be careful, it would seem that you have pushed (again) your branch to the main repo. Could you remove it please?

**Comment by jfehrle** [comment]
@Zimmi48 No, I've not pushed to the main repo.  Read the PR.  This PR is in my repo.  The incorrect one #10483 was closed and "deleted" on July 5.  Not sure why it still shows up.

image

image

image

**Comment by hendriktews** [comment]
@jfehre: Jim, I have prooftree working now with this PR and with PR 439 of Proofgeneral/PG. There are still a number of issues that I need to fix in prooftree before making a release, but it is working fine and I have used it quite a bit already. Thanks again for the effort you put into this!

File: engine/evd.mli
Comment (mattam82) [inline engine/evd.mli]: This looks incorrect, it includes *defined* evars

**Comment by jfehrle** [comment]
@Zimmi48 I just created a `dependent_evars` branch in master by mistake.  Not sure how to delete it.  I still had an incorrect remote specification for that source tree, which I just fixed.  Thanks.

**Comment by Zimmi48** [comment]
I just deleted it (it can be done in GitHub's web interface).

**Comment by jfehrle** [comment]
@mattam82 I updated the comment.  Please proceed.  Thanks.

**Comment by jfehrle** [comment]
I think the script failures are irrelevant.

**Comment by jfehrle** [comment]
@herbelin, I've been advised that you are best candidate to assign this PR to.  Can you take a final look at it and, if agreeable, merge it?  The PR has been waiting since the beginning of August.  I was under the impression the @ejgallego would take care of it, but he seems not to have time to do so.  I would like to get this in for 10.11, which means it needs to be merged by November 1.  Please let me know your thoughts.  Thanks.

**Review by herbelin — APPROVED** [review]
This is good to me. (In particular, answering to a former comment, I approve the name `evar_nodes_of_term`.) 

Being approved in parallel by @Zimmi48 and @ejgallego, and also having waited for two months for presumably no other reasons than the absence of a dedicated person to occupy the role of ensuring that no PRs are stalling, I'm proposing to merge by the end of the week (@mattam82, shouldn't we definitely organize ourselves so that such a role is filled, even if, for instance, we fill it in alternance?).

**Comment by jfehrle** [comment]
Great, thanks @herbelin!

**Comment by hendriktews** [comment]
@jfehrle : Thanks again for the effort you put into this. I am happily using the patch for several weeks already to have prooftree visualization for my proofs in Coq.

**Comment by jfehrle** [comment]
Sure thing.  I hope you'll publicize the tool once 8.11 comes out.  Also worthwhile to seek out feedback on the tool--others may have good ideas or find issues.  If you think the tool may be useful in teaching Coq, maybe contact some people who have taught Coq, such as @Zimmi48 or @erikmd.

**Pull request merged**
